### PR TITLE
fix: force xgrammar backend in from_sampling_params

### PR DESCRIPTION
**Pull request by JMPSequeira**
## Problem

In the b12x fork, `SamplingParams.validate()` — which auto-selects the structured output backend via `_validate_structured_outputs` — may not be called before `Request` creation. When this happens, `StructuredOutputsParams._backend` stays `None`, and the scheduler skips grammar enforcement entirely. The `structured_outputs` field is accepted by the API, the grammar is validated and compiled, but the bitmask is never applied to logits during sampling. **Constraints are silently ignored.**

## Reproduction

Send a completion request with a grammar constraint:

```json
{
  "model": "GLM-5.2",
  "prompt": "Say: ",
  "max_tokens": 5,
  "structured_outputs": {"grammar": "root ::= \"hello\""}
}
```

On the unpatched b12x image, the model outputs freely (e.g. `"1. Who they are"`) instead of being constrained to `"hello"`.

## Fix

Force the xgrammar backend in `from_sampling_params()` when `validate()` hasn't already selected one. This is a no-op when `validate()` has run (the backend is already set).

## Verification

Tested on the eldritch-fullstack image (`v0.11.2.dev279`):

1. **Simple grammar**: `root ::= "hello"` — output constrained to `"hello"` ✓
2. **Full LLang GBNF grammar**: 75 nonterminals, 154K-token vocab (GLM-5.2 tokenizer) — output constrained to valid LLang syntax ✓
3. **Model**: Qwen3.5-9B-NVFP4 — generated valid LLang code (pattern matching, `let-in`, pipeline) under constraint ✓

## Notes

- This issue was diagnosed and tested using **GLM-5.2 (504B)** as the analysis and reasoning model.
- The fix is 8 lines, all additive — no existing behavior is changed when `validate()` has already run.
- The `structured_outputs` API field, grammar validation, xgrammar compilation, and model runner bitmask application all work correctly in isolation. The only gap was the backend selection not happening before `Request` creation.

## Summary by CodeRabbit

* **Bug Fixes**
  * Structured outputs now more reliably use the server’s configured backend when a request does not include one.
  * Requests that rely on automatic backend selection will now consistently choose the preferred compatible backend when available, reducing setup-related failures.

### CODE REVIEW

File: vllm/v1/structured_output/request.py [hunk 1]
```diff
@@ -37,6 +37,14 @@ def from_sampling_params(
...
         if not params or params.all_constraints_none():
             return None
+        # Force the xgrammar backend if validate() hasn't already selected one.
+        # In the b12x fork, SamplingParams.validate() (which auto-selects the
+        # backend via _validate_structured_outputs) may not be called before
+        # Request creation. Without a backend, the scheduler skips grammar
+        # enforcement and constraints are silently ignored.
+        if not params._backend:
+            params._backend = "xgrammar"
+            params._backend_was_auto = True
```
[the remark this answers is not in the document]
Comment (JMPSequeira) [reply vllm/v1/structured_output/request.py]: You're right — hard-coding `xgrammar` here was wrong. I've reverted the `request.py` change entirely and moved the fix to `StructuredOutputManager` in `__init__.py`, where `vllm_config` is available. The fallback now uses the server-configured backend (`vllm_config.structured_outputs_config.backend`) and only resolves `"auto"` to `"xgrammar"` as a last resort. This respects operator-configured backends and doesn't mutate `params._backend`.

File: vllm/v1/structured_output/__init__.py [hunk 2]
```diff
@@ -130,6 +130,16 @@ def grammar_init(self, request: "Request") -> None:
...
+            if not backend:
+                # Fallback: use the server-configured backend when
+                # SamplingParams.validate() hasn't auto-selected one.
+                # In the b12x fork, validate() may not be called before
+                # Request creation, leaving _backend as None.
+                backend = self.vllm_config.structured_outputs_config.backend
+            if backend == "auto":
+                # Resolve auto to xgrammar (the preferred backend when
+                # xgrammar is installed, which it is in vLLM by default).
+                backend = "xgrammar"
```
[the remark this answers is not in the document]
Comment (JMPSequeira) [reply vllm/v1/structured_output/__init__.py]: Fixed. The `auto` case now tries xgrammar first and falls back to guidance/outlines, matching the resolution order in `SamplingParams.validate()` (lines 967-1006). No backend is hard-coded — the fallback chain is: try `XgrammarBackend()`, catch and try `GuidanceBackend()`, catch and try `OutlinesBackend()`. This preserves the existing auto-resolution contract.

**Comment by malaiwah** [comment]
## Superseded by PR #59 — bug reproduction confirmed

PR #59 confirms this PR's hypothesis was incorrect: the  selection was not the root cause. The actual bug is the reasoning gate in  /  — see the reproduction evidence on PR #59: https://github.com/local-inference-lab/vllm/pull/59#issuecomment-5223532763

**Bug reproduced**: completions requests with  configured silently ignore grammar constraints (output:  instead of ). Chat requests with  correctly enforce grammar (output: ).

This PR can be closed in favor of #59.

**Comment by malaiwah** [comment]
## Superseded by PR #59 — bug reproduction confirmed

PR #59 confirms this PR's hypothesis was incorrect: the `_backend` selection was not the root cause. The actual bug is the reasoning gate in `StructuredOutputManager.should_fill_bitmask` / `should_advance` — see the reproduction evidence on PR #59: https://github.com/local-inference-lab/vllm/pull/59#issuecomment-5223532763

**Bug reproduced**: completions requests with `--reasoning-parser glm45` configured silently ignore grammar constraints (output: "1, 2, 3, 4" instead of "hello"). Chat requests with `enable_thinking=false` correctly enforce grammar (output: "hello").

This PR can be closed in favor of #59.
